sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Main.scala

```diff
@@ -36,55 +35,54 @@ object Main extends IOApp {
 
-    val eventStream: Stream[F, (collector.CollectorPayload, List[Event])] = {
+    val eventStream: Stream[F, GenOutput] = {
       config.duplicates match {
         case Some(dups) => Stream.repeatEval(Sync[F].delay(
           runGen(SdkEvent.genPairDup(dups.natProb, dups.synProb, dups.natTotal, dups.synTotal, config.eventPerPayloadMin, config.eventPerPayloadMax), rng)))
         case None => Stream.repeatEval(Sync[F].delay(runGen(SdkEvent.genPair(config.eventPerPayloadMin, config.eventPerPayloadMax), rng)))
       }
     }
 
-    def sinkBuilder(prefix: String) = if (outputDir.toString.startsWith("file:")) {
-      RotatingSink.file(prefix, outputDir, config.payloadsPerFile)
-    } else if (outputDir.toString.startsWith("s3:")) {
-      RotatingSink.s3(prefix, outputDir, config.payloadsPerFile)
-    } else {
-      throw new RuntimeException(s"Unknown scheme in $outputDir")
-    }
+    def sinkBuilder(prefix: String, idx: Int): Pipe[F, Byte, INothing] =
+      if (outputDir.toString.startsWith("file:")) {
+        RotatingSink.file(prefix, idx, outputDir)
+      } else if (outputDir.toString.startsWith("s3:")) {
+        RotatingSink.s3(prefix, idx, outputDir)
+      } else {
+        throw new RuntimeException(s"Unknown scheme in $outputDir")
+      }
 
-    val rawSink: Resource[F, RotatingSink[F]] = if (config.withRaw)
-      sinkBuilder("raw")
-    else
-      RotatingSink.void
+    def rawSink(idx: Int): Pipe[F, Byte, INothing] =
+      if (config.withRaw)
+        sinkBuilder("raw", idx)
+      else
+        _.drain
 
-    val enrichSink: Resource[F, RotatingSink[F]] = if (config.withRaw)
+    def enrichSink(idx: Int): Pipe[F, Byte, INothing] =
       config.enrichFormat match {
-        case EnrichFormat.Json => sinkBuilder("json-enrich")
-        case EnrichFormat.Tsv => sinkBuilder("tsv-enrich")
+        case EnrichFormat.Json => sinkBuilder("json-enrich", idx)
+        case EnrichFormat.Tsv => sinkBuilder("tsv-enrich", idx)
       }
-    else RotatingSink.void
 
-
-    (for {
-      raw <- rawSink
-      enrich <- enrichSink
-    } yield (raw, enrich)).use { case (raw, enrich) =>
-      eventStream
-        .zipWithIndex
-        .takeWhile(_._2 < config.payloadsTotal)
-        .map(_._1)
-        .evalTap {
-          case (payload, sdkEvents) =>
-            raw.write(Stream.emit(payload).through(Serializers.rawSerializer(config.compress))) *>
-              enrich.write(Stream.emits(sdkEvents).through(Serializers.enrichedSerializer(config.enrichFormat, config.compress)))
-        }
-        .compile
-        .drain
-    }
+    eventStream
+      .take(config.payloadsTotal.toLong)
+      .through(RotatingSink.rotate(config.payloadsPerFile) { idx =>
+        val pipe1: Pipe[F, GenOutput, INothing]  = _.map(_._1)
+          .through(Serializers.rawSerializer(config.compress))
+          .through(rawSink(idx))
+        val pipe2: Pipe[F, GenOutput, INothing]  = _.flatMap(in => Stream.emits(in._2))
```

Comment:
It does.  This was deliberate at the time.

I arranged it so that the payloads in raw-0042 correspond exactly to the events in enriched-0042 and in transformed-0042.  Even though this might mean that the `enriched-0042` has more lines than the `raw-0042` file.

The alternative is to set things up so that each raw file has strictly `n` lines, and each enriched file also has strictly `n` rows, but it loses the correspondence between matching files.

It's a close call which is more preferable.  I slightly prefer the current set up with better correspondence between output files.  I would welcome your opinion on this.

The configuration option is called `payloadsPerFile` so I don't think it breaks a contract to exceed the number of _events_ in the file.